### PR TITLE
Reduce ridge filters memory footprints

### LINKED ISSUE #6507: Sato filter memory usage with multiple sigmas
## Description

Sato filter and other ridge filters in _ridges.py_ have quite large unnecessary memory usage if called with multiple sigma values. It is basically allocating for all results, storing results and then taking maximum, while it is possible to just take max after each sigma.

Original code:
```python
    # Generate empty (n+1)D arrays for storing auxiliary images filtered
    # at different (sigma) scales
    filtered_array = np.zeros(sigmas.shape + image.shape, dtype=float_dtype)

    # Filtering for all (sigma) scales
    for i, sigma in enumerate(sigmas):

        # Calculate (sorted) eigenvalues
        lamba1, *lambdas = compute_hessian_eigenvalues(image, sigma,
                                                       sorting='val',
                                                       mode=mode, cval=cval)

        # Compute tubeness, see  equation (9) in reference [1]_.
        # np.abs(lambda2) in 2D, np.sqrt(np.abs(lambda2 * lambda3)) in 3D
        filtered = np.abs(np.multiply.reduce(lambdas)) ** (1/len(lambdas))

        # Remove background and store results in (n+1)D matrices
        filtered_array[i] = np.where(lambdas[-1] > 0, filtered, 0)

    # Return for every pixel the maximum value over all (sigma) scales
    return np.max(filtered_array, axis=0)
```

Modified code:
```python
    # Generate empty (n+1)D arrays for storing auxiliary images filtered
    # at different (sigma) scales
    filtered_array = np.zeros(image.shape, dtype=float_dtype)

    # Filtering for all (sigma) scales
    for i, sigma in enumerate(sigmas):

        # Calculate (sorted) eigenvalues
        lamba1, *lambdas = compute_hessian_eigenvalues(image, sigma,
                                                       sorting='val',
                                                       mode=mode, cval=cval)

        # Compute tubeness, see  equation (9) in reference [1]_.
        # np.abs(lambda2) in 2D, np.sqrt(np.abs(lambda2 * lambda3)) in 3D
        filtered = np.abs(np.multiply.reduce(lambdas)) ** (1/len(lambdas))

        # Remove background and store results in (n+1)D matrices
        filtered = np.where(lambdas[-1] > 0, filtered, 0)
        filtered_array = np.maximum(filtered_array,filtered)

    # Return for every pixel the maximum value over all (sigma) scales
    return filtered_array
```

## Way to reproduce
Call sato filter with huge image and multiple sigmas and observe memory use.

## Version information
`
3.10.4 | packaged by conda-forge | (main, Mar 30 2022, 08:38:02) [MSC v.1916 64 bit (AMD64)]
Windows-10-10.0.19043-SP0
scikit-image version: 0.19.2
numpy version: 1.21.5
`

### PR DESCRIPTION
**Pull request by tkumpumaki**
## Description
Reduce ridge filters memory footprints by not storing intermediate results for final max computation. Instead after each sigma take max between last and current sigma result.

Closes #6507

## For reviewers

- Check that the PR title is short, concise, and will make sense 1 year
  later.
- Check that new functions are imported in corresponding `__init__.py`.
- Check that new features, API changes, and deprecations are mentioned in
  `doc/release/release_dev.rst`.
- There is a bot to help automate backporting a PR to an older branch. For
  example, to backport to v0.19.x after merging, add the following in a PR
  comment: `@meeseeksdev backport to v0.19.x`
- To run benchmarks on a PR, add the `run-benchmark` label. To rerun, the label
  can be removed and then added again. The benchmark output can be checked in
  the "Actions" tab.

### CODE REVIEW

**Review by mkcor — APPROVED** [review]
@tkumpumaki thank you so much for your contribution!

**Comment by mkcor** [comment]
@scikit-image/core since memory use is greatly improved with this PR, should we include a benchmark?

File: skimage/filters/ridges.py [hunk 1]
```diff
@@ -456,17 +458,18 @@ def frangi(image, sigmas=range(1, 10, 2), scale_range=None,
 
         # Compute output image for given (sigma) scale and store results in
         # (n+1)D matrices, see equations (13) and (15) in reference [1]_
-        filtered_array[i] = ((1 - np.exp(-r_a / alpha_sq))
+        filtered = ((1 - np.exp(-r_a / alpha_sq))
                              * np.exp(-r_b / beta_sq)
                              * (1 - np.exp(-r_g / gamma_sq)))
```
Comment (rfezzani) [inline skimage/filters/ridges.py:463]:
```suggestion
        filtered = ((1 - np.exp(-r_a / alpha_sq))
                     * np.exp(-r_b / beta_sq)
                     * (1 - np.exp(-r_g / gamma_sq)))
```

**Review by rfezzani — APPROVED** [review]
Excellent! Thank you @tkumpumaki :wink: 
I simply suggest a small code formatting, otherwise every think seems fine. @mkcor, not sure if a benchmark is necessary, but up to other @scikit-image/core members :slightly_smiling_face:

**Comment by lagru** [comment]
> not sure if a benchmark is necessary, but up to other https://github.com/orgs/scikit-image/teams/core members 

I'd say yes. @tkumpumaki let us know if you want to do that yourself or we will happily do so for you if you are not that familiar with asv.

**Comment by tkumpumaki** [comment]
> > not sure if a benchmark is necessary, but up to other https://github.com/orgs/scikit-image/teams/core members
> 
> I'd say yes. @tkumpumaki let us know if you want to do that yourself or we will happily do so for you if you are not that familiar with asv.

I don't have a slightest clue how to test memory usage with the asv as python is not my main weapon. ;)
I would guess that CPU time is not that different as same amount of computation is done.

**Comment by grlee77** [comment]
> I don't have a slightest clue how to test memory usage with the asv as python is not my main weapon. ;)

no worries @tkumpumaki. @lagru, are you interested in adding the benchmark? If so that is great and we can wait for that, but if no one plans to work on it soon let's not hold up the PR over it.

Certainly the memory usage will be lower with this change and I doubt there is a substantial change in runtime. There will be more calls to maximum, but each one does less work and the maximum operations are not the most expensive operations in this function in any case. Even if there were a small performance decrease, I would still vote to merge this for the clear memory efficiency gain!

**Comment by lagru** [comment]
@grlee77 happy to. I'll do it in a minute.

**Comment by lagru** [comment]
<details>
<summary>Ony my machine</summary>

```
$ asv continuous main 7535320f661c2734321edbfe24f9fb3e5ae3bc6f -b RidgeFilters
· Creating environments
· Discovering benchmarks
·· Uninstalling from virtualenv-py3.10-cython-numpy1.23-pooch-pythran-scipy-virtualenv
·· Installing 7535320f <ridge-operator-less-memory-use~1> into virtualenv-py3.10-cython-numpy1.23-pooch-pythran-scipy-virtualenv.
· Running 10 total benchmarks (2 commits * 1 environments * 5 benchmarks)
[  0.00%] · For scikit-image commit 7535320f <ridge-operator-less-memory-use~1> (round 1/1):
[  0.00%] ·· Benchmarking virtualenv-py3.10-cython-numpy1.23-pooch-pythran-scipy-virtualenv
[ 10.00%] ··· benchmark_filters.RidgeFilters.peakmem_frangi                                        394M
[ 20.00%] ··· benchmark_filters.RidgeFilters.peakmem_hessian                                       394M
[ 30.00%] ··· benchmark_filters.RidgeFilters.peakmem_meijering                                     380M
[ 40.00%] ··· benchmark_filters.RidgeFilters.peakmem_sato                                          269M
[ 50.00%] ··· benchmark_filters.RidgeFilters.peakmem_setup                                         143M
[ 50.00%] · For scikit-image commit 6342ec59 <main> (round 1/1):
[ 50.00%] ·· Building for virtualenv-py3.10-cython-numpy1.23-pooch-pythran-scipy-virtualenv.
[ 50.00%] ·· Benchmarking virtualenv-py3.10-cython-numpy1.23-pooch-pythran-scipy-virtualenv
[ 60.00%] ··· benchmark_filters.RidgeFilters.peakmem_frangi                                        528M
[ 70.00%] ··· benchmark_filters.RidgeFilters.peakmem_hessian                                       523M
[ 80.00%] ··· benchmark_filters.RidgeFilters.peakmem_meijering                                     412M
[ 90.00%] ··· benchmark_filters.RidgeFilters.peakmem_sato                                          332M
[100.00%] ··· benchmark_filters.RidgeFilters.peakmem_setup                                         145M
before           after         ratio
[6342ec59]       [7535320f]
<main>           <ridge-operator-less-memory-use~1>
-            332M             269M     0.81  benchmark_filters.RidgeFilters.peakmem_sato
-            523M             394M     0.75  benchmark_filters.RidgeFilters.peakmem_hessian
-            528M             394M     0.75  benchmark_filters.RidgeFilters.peakmem_frangi

SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
PERFORMANCE INCREASED.
```

</details>

**Comment by lagru** [comment]
@tkumpumaki, I pushed to your branch. To avoid conflicts, I recommend pulling these changes if and before you add commits yourself again: `git pull origin ridge-operator-less-memory-use` (assuming your fork is associated with `origin`).

**Review by lagru — APPROVED** [review]

**Comment by grlee77** [comment]
The stochastic RANSAC failure on CI is unrelated.

**Comment by grlee77** [comment]
Thanks for the quick update, @lagru . While we are adding the benchmarks, can you also add `time_meijering`, etc. as well?

**Comment by lagru** [comment]
It seems that the label `run-benchmarks` triggers something but the workflow is shown as "Skipped" nevertheless. Does someone know why?

**Comment by lagru** [comment]
<details>
<summary>My machine, with CPU benchmarks added</summary>

```
$ asv continuous main HEAD -b RidgeFilters
· Creating environments
· Discovering benchmarks
·· Uninstalling from virtualenv-py3.10-cython-numpy1.23-pooch-pythran-scipy-virtualenv
·· Installing 1d0e9d37 <pr6509-ridge-operator-less-memory-use> into virtualenv-py3.10-cython-numpy1.23-pooch-pythran-scipy-virtualenv.
· Running 18 total benchmarks (2 commits * 1 environments * 9 benchmarks)
[  0.00%] · For scikit-image commit bfcb6298 <main> (round 1/2):
[  0.00%] ·· Building for virtualenv-py3.10-cython-numpy1.23-pooch-pythran-scipy-virtualenv..
[  0.00%] ·· Benchmarking virtualenv-py3.10-cython-numpy1.23-pooch-pythran-scipy-virtualenv
[ 16.67%] ··· Running (benchmark_filters.RidgeFilters.time_frangi--)...
[ 25.00%] ··· Running (benchmark_filters.RidgeFilters.time_sato--).
[ 25.00%] · For scikit-image commit 1d0e9d37 <pr6509-ridge-operator-less-memory-use> (round 1/2):
[ 25.00%] ·· Building for virtualenv-py3.10-cython-numpy1.23-pooch-pythran-scipy-virtualenv.
[ 25.00%] ·· Benchmarking virtualenv-py3.10-cython-numpy1.23-pooch-pythran-scipy-virtualenv
[ 41.67%] ··· Running (benchmark_filters.RidgeFilters.time_frangi--)...
[ 50.00%] ··· Running (benchmark_filters.RidgeFilters.time_sato--).
[ 50.00%] · For scikit-image commit 1d0e9d37 <pr6509-ridge-operator-less-memory-use> (round 2/2):
[ 50.00%] ·· Benchmarking virtualenv-py3.10-cython-numpy1.23-pooch-pythran-scipy-virtualenv
[ 52.78%] ··· benchmark_filters.RidgeFilters.peakmem_frangi                                        394M
[ 55.56%] ··· benchmark_filters.RidgeFilters.peakmem_hessian                                       394M
[ 58.33%] ··· benchmark_filters.RidgeFilters.peakmem_meijering                                     380M
[ 61.11%] ··· benchmark_filters.RidgeFilters.peakmem_sato                                          268M
[ 63.89%] ··· benchmark_filters.RidgeFilters.peakmem_setup                                         143M
[ 66.67%] ··· benchmark_filters.RidgeFilters.time_frangi                                        1.35±0s
[ 69.44%] ··· benchmark_filters.RidgeFilters.time_hessian                                       1.36±0s
[ 72.22%] ··· benchmark_filters.RidgeFilters.time_meijering                                     4.89±0s
[ 75.00%] ··· benchmark_filters.RidgeFilters.time_sato                                          935±7ms
[ 75.00%] · For scikit-image commit bfcb6298 <main> (round 2/2):
[ 75.00%] ·· Building for virtualenv-py3.10-cython-numpy1.23-pooch-pythran-scipy-virtualenv.
[ 75.00%] ·· Benchmarking virtualenv-py3.10-cython-numpy1.23-pooch-pythran-scipy-virtualenv
[ 77.78%] ··· benchmark_filters.RidgeFilters.peakmem_frangi                                        527M
[ 80.56%] ··· benchmark_filters.RidgeFilters.peakmem_hessian                                       527M
[ 83.33%] ··· benchmark_filters.RidgeFilters.peakmem_meijering                                     448M
[ 86.11%] ··· benchmark_filters.RidgeFilters.peakmem_sato                                          336M
[ 88.89%] ··· benchmark_filters.RidgeFilters.peakmem_setup                                         145M
[ 91.67%] ··· benchmark_filters.RidgeFilters.time_frangi                                        1.37±0s
[ 94.44%] ··· benchmark_filters.RidgeFilters.time_hessian                                       1.37±0s
[ 97.22%] ··· benchmark_filters.RidgeFilters.time_meijering                                  4.88±0.01s
[100.00%] ··· benchmark_filters.RidgeFilters.time_sato                                          941±4ms
before           after         ratio
[bfcb6298]       [1d0e9d37]
<main>           <pr6509-ridge-operator-less-memory-use>
-            448M             380M     0.85  benchmark_filters.RidgeFilters.peakmem_meijering
-            336M             268M     0.80  benchmark_filters.RidgeFilters.peakmem_sato
-            527M             394M     0.75  benchmark_filters.RidgeFilters.peakmem_hessian
-            527M             394M     0.75  benchmark_filters.RidgeFilters.peakmem_frangi

SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
PERFORMANCE INCREASED.
```
</details>

**Comment by grlee77** [comment]
Hi @tkumpumaki, I merged a larger refactor of ridge filters by @anntzer via #6446 today. It has created conflicts here, but more importantly it had also refactored these same sections and removed the upfront memory allocations! 

Given that, it may be that no code changes are necessary in `ridge.py` for this PR? In that case, please just update the issue title to something like "add ridge filter benchmarks".

My apologies for not realizing the full overlap between these two and pointing it out earlier. Let us know if you need help resolving conflicts here.

**Comment by anntzer** [comment]
That PR removed the large upfront memory allocations but still first computes the filters for all sigmas before taking the max (so there's a large memory use just before the max is taken); it may still be useful to replace that by computing the pairwise maxes after each sigma computation.

**Comment by tkumpumaki** [comment]
I merged main branch back to this branch and replaced _ridges.py_ with the new one from the main. Then added memory fix again.

**Comment by lagru** [comment]
I'm kind of confused by the [commit hash that is used by the benchmark workflow](https://github.com/scikit-image/scikit-image/actions/runs/3084426412/jobs/4986564711#step:7:75) which is d3e5df65651ed386dcd41a7da7b94cee8af12cef. It looks like a merge commit but is not listed on this PR...

**Comment by lagru** [comment]
But it shows an decreased memory use locally:

<details>
<summary>Click to expand</summary>

```
$ asv continuous --show-stderr --split main pr6509-ridge-operator-less-memory-use -b RidgeFilters
· Creating environments
· Discovering benchmarks
·· Uninstalling from virtualenv-py3.10-cython-numpy1.23-pooch-pythran-scipy-virtualenv
·· Installing 4a792387 <pr6509-ridge-operator-less-memory-use> into virtualenv-py3.10-cython-numpy1.23-pooch-pythran-scipy-virtualenv.
· Running 18 total benchmarks (2 commits * 1 environments * 9 benchmarks)
[  0.00%] · For scikit-image commit daa991ad <main> (round 1/2):
[  0.00%] ·· Building for virtualenv-py3.10-cython-numpy1.23-pooch-pythran-scipy-virtualenv................................................................................
[  0.00%] ·· Benchmarking virtualenv-py3.10-cython-numpy1.23-pooch-pythran-scipy-virtualenv
[ 16.67%] ··· Running (benchmark_filters.RidgeFilters.time_frangi--)...
[ 25.00%] ··· Running (benchmark_filters.RidgeFilters.time_sato--).
[ 25.00%] · For scikit-image commit 4a792387 <pr6509-ridge-operator-less-memory-use> (round 1/2):
[ 25.00%] ·· Building for virtualenv-py3.10-cython-numpy1.23-pooch-pythran-scipy-virtualenv.
[ 25.00%] ·· Benchmarking virtualenv-py3.10-cython-numpy1.23-pooch-pythran-scipy-virtualenv
[ 41.67%] ··· Running (benchmark_filters.RidgeFilters.time_frangi--)...
[ 50.00%] ··· Running (benchmark_filters.RidgeFilters.time_sato--).
[ 50.00%] · For scikit-image commit 4a792387 <pr6509-ridge-operator-less-memory-use> (round 2/2):
[ 50.00%] ·· Benchmarking virtualenv-py3.10-cython-numpy1.23-pooch-pythran-scipy-virtualenv
[ 52.78%] ··· benchmark_filters.RidgeFilters.peakmem_frangi                                        332M
[ 55.56%] ··· benchmark_filters.RidgeFilters.peakmem_hessian                                       332M
[ 58.33%] ··· benchmark_filters.RidgeFilters.peakmem_meijering                                     279M
[ 61.11%] ··· benchmark_filters.RidgeFilters.peakmem_sato                                          268M
[ 63.89%] ··· benchmark_filters.RidgeFilters.peakmem_setup                                         145M
[ 66.67%] ··· benchmark_filters.RidgeFilters.time_frangi                                      5.49±0.2s
[ 69.44%] ··· benchmark_filters.RidgeFilters.time_hessian                                    5.58±0.05s
[ 72.22%] ··· benchmark_filters.RidgeFilters.time_meijering                                  5.18±0.04s
[ 75.00%] ··· benchmark_filters.RidgeFilters.time_sato                                       4.80±0.03s
[ 75.00%] · For scikit-image commit daa991ad <main> (round 2/2):
[ 75.00%] ·· Building for virtualenv-py3.10-cython-numpy1.23-pooch-pythran-scipy-virtualenv.
[ 75.00%] ·· Benchmarking virtualenv-py3.10-cython-numpy1.23-pooch-pythran-scipy-virtualenv
[ 77.78%] ··· benchmark_filters.RidgeFilters.peakmem_frangi                                        382M
[ 80.56%] ··· benchmark_filters.RidgeFilters.peakmem_hessian                                       382M
[ 83.33%] ··· benchmark_filters.RidgeFilters.peakmem_meijering                                     329M
[ 86.11%] ··· benchmark_filters.RidgeFilters.peakmem_sato                                          318M
[ 88.89%] ··· benchmark_filters.RidgeFilters.peakmem_setup                                         143M
[ 91.67%] ··· benchmark_filters.RidgeFilters.time_frangi                                     5.14±0.03s
[ 94.44%] ··· benchmark_filters.RidgeFilters.time_hessian                                    5.35±0.09s
[ 97.22%] ··· benchmark_filters.RidgeFilters.time_meijering                                  5.13±0.02s
[100.00%] ··· benchmark_filters.RidgeFilters.time_sato                                        4.55±0.1s
before           after         ratio
[daa991ad]       [4a792387]
<main>           <pr6509-ridge-operator-less-memory-use>
-            382M             332M     0.87  benchmark_filters.RidgeFilters.peakmem_frangi
-            382M             332M     0.87  benchmark_filters.RidgeFilters.peakmem_hessian
-            329M             279M     0.85  benchmark_filters.RidgeFilters.peakmem_meijering
-            318M             268M     0.84  benchmark_filters.RidgeFilters.peakmem_sato

SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
PERFORMANCE INCREASED.
```

</details>

**Review by lagru — APPROVED** [review]

File: skimage/filters/ridges.py [hunk 2]
```diff
@@ -149,7 +149,9 @@ def meijering(image, sigmas=range(1, 10, 2), alpha=None,
     mtx = linalg.circulant(
         [1, *[alpha] * (image.ndim - 1)]).astype(image.dtype)
 
-    filtered = []
+    # Generate empty array for storing maximum value
+    # from different (sigma) scales
+    filtered_max = np.zeros(image.shape, image.dtype)
```
Comment (rfezzani) [inline skimage/filters/ridges.py:154]:
```suggestion
    filtered_max = np.zeros_like(image)
```

File: skimage/filters/ridges.py [hunk 3]
```diff
@@ -221,7 +224,9 @@ def sato(image, sigmas=range(1, 10, 2), black_ridges=True,
     if not black_ridges:  # Normalize to black ridges.
         image = -image
 
-    filtered = []
+    # Generate empty array for storing maximum value
+    # from different (sigma) scales
+    filtered_max = np.zeros(image.shape, image.dtype)
```
Comment (rfezzani) [inline skimage/filters/ridges.py:229]:
```suggestion
    filtered_max = np.zeros_like(image)
```

File: skimage/filters/ridges.py [hunk 4]
```diff
@@ -318,7 +323,9 @@ def frangi(image, sigmas=range(1, 10, 2), scale_range=None,
...
```
Comment (rfezzani) [inline skimage/filters/ridges.py:328]:
```suggestion
    filtered_max = np.zeros_like(image)
```

**Review by rfezzani — APPROVED** [review]
LGTM, simply recommending the use of `np.zeros_like` when possible.

**Comment by lagru** [comment]
Thanks everyone!

**Comment by mkcor** [comment]
> I'm kind of confused by the [commit hash that is used by the benchmark workflow](https://github.com/scikit-image/scikit-image/actions/runs/3084426412/jobs/4986564711#step:7:75) which is [d3e5df6](https://github.com/scikit-image/scikit-image/commit/d3e5df65651ed386dcd41a7da7b94cee8af12cef). It looks like a merge commit but is not listed on this PR...

Right...  It looks as if CI, instead of running directly for the latest commit (i.e., [4a79238](https://github.com/scikit-image/scikit-image/pull/6509/commits/4a79238745ff41b96a74405b39ed4780ba60056c)), would merge this very commit into an *ad hoc* branch before running it.

**Pull request merged**